### DISCUSSION TITLE
File-based configuration

### DISCUSSION THREAD

**Discussion by wsw70** [alexflint/go-arg#274] [Ideas]
TL;DR: It would be awesome if it was possible to write and read a configuration

I am writing a CLI with several defaults that fit my case. Others who will (hopefully) use it may have different defaults (this is going to be an open-source CLI so the configurations may be really different).

The solution is to use very long command lines, which is impractical because this is expected to be a CLI one uses often. To fix this, one can create shortcuts in their shell but this is a workaround.

A nice solution would be to read the configuration from a file, and then call `mustParse` to replace arguments that may have been added on the command line. In other words, the fallback would be

- command line arguments
- configuration file
- defaults

I understand that this is at the edge of a "command line parser" but with some eye squinting it could fit, no? :)

**Comment by alexflint** [comment]
Definitely. Have you tried loading a config file into the argument structure using an external library, then just running arg.MustParse over that? Something like:

```go
var args struct {
  Test string `arg:"--test" json:"test"`
}
buf, _ := os.ReadFile(configFilePath)
_ = json.Unmarshal(buf, &args)
arg.MustParse(&args)
```

There are several ways that this isn't quite perfect, but I wonder if something like this would work for your use case.

**Reply by wsw70** [reply]
Thank you for your answer. Yes, I had this idea as well - the "problem" is that there is just one pass on `MustParse()` - which means that the command line arguments would not be considered.

I was rather thinking about something like

- defining the `args` structure like usual
- reading in a configuration file (in JSON for instance) → this would populate the fields from `args` that are present in the JSON
- parsing the actual arguments → this would add or overwrite existing `args` (add when there are none, and update if some are already set up via the config file or defaults)

If I get what you wrote right, it would be rather "replace command line arguments with a configuration file" (?)
